### PR TITLE
Infinite random-file random-line stateless sampler

### PR DESCRIPTION
**Pull request by pzelasko**
As proposed in #1096 

It's pretty much what was described, with at most minor deviations. @danpovey let me know what you think, you can find the usage in unit tests. I couldn't think of a name that would describe what this sampler does, so I initially called it `PoveySampler` -- we should continue to build on the legacy of the Povey feature window :)

### CODE REVIEW

File: lhotse/dataset/sampling/povey.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+from lhotse.dataset import DynamicBucketingSampler, DynamicCutSampler
+from lhotse.dataset.sampling.base import SamplingDiagnostics
+from lhotse.lazy import Dillable
+from lhotse.serialization import decode_json_line
+from lhotse.utils import Pathlike
+
+PathlikeAndScale = Tuple[Pathlike, float]
+
+
+class PoveySampler(torch.utils.data.Sampler, Dillable):
```
Comment (csukuangfj) [inline lhotse/dataset/sampling/povey.py:21]: There is a "povey window". 
Now we also have a "povey sampler"  😀

Comment (danpovey) [reply lhotse/dataset/sampling/povey.py:21]: LOL, but let's call it StatelessSampler.  With the Povey Window, I think people didn't get that it was basically a joke (since no-one really cares about things like windows these days), and they perhaps thought that it was a kind of self-aggrandizement.

File: lhotse/dataset/sampling/povey.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+        >>> sampler = PoveySampler(
+        ...     cuts_paths=["data/cuts_a.jsonl", "data/cuts_b.jsonl"],
+        ...     index_path="data/files.idx",
+        ...     max_duration=600.0,
+        ... )
+
+
+    .. note:: This sampler works only with uncompressed jsonl manifests, as it creates extra index files with line byte offsets to quickly find and sample JSON lines.
+     This means this sampler will not work with Webdataset and Lhotse Shar data format.
```
Comment (danpovey) [inline lhotse/dataset/sampling/povey.py:53]: Perhaps we should mention that if you restart training, you should set the random seed to some function of the step you restarted at, to ensure that you don't see the exact same data that you saw at the beginning of training (or at some previous restart point).

Comment (pzelasko) [reply lhotse/dataset/sampling/povey.py:53]: Good point, I'll also add some logging that prints out the random seeds and maybe other info at initialization so it'll be possible to keep track of these things between experiments.

**Comment by pzelasko** [comment]
Check it now, I changed it to use OS TRNG for sampling the base seed (if available) so that it doesn't depend on Python's global seed at all.

**Comment by pzelasko** [comment]
I'll merge it, let me know later if it works as you expected.

**Comment by danpovey** [comment]
I'm not sure about the decisions about the TRNG stuff.  My plan was to have the base seed depend on the python RNG state.  The idea was that in the main thread, we set the RNG state to be a function of the start-iter if you are restarting, so if you are restarting training you get a different random seed.  OTOH it would be consistent between runs, so that if you get a crash it can be debugged.  Otherwise debugging crashes would be quite hard!

**Comment by pzelasko** [comment]
It's going to print the seed it chose in the logs, and if you want to reproduce the run, you can pass the argument `base_seed=<value>` to the sampler (by default it's None which means choose it at random). The same argument can be used to pass another seed of your own choice.

**Comment by danpovey** [comment]
Hm, OK.  When we use this I plan to pass in something like the start iteration then.
If the secrets module gives a different number per worker, it seems to me it would be quite painful to pass
in separate numbers per worker in order to reproduce a crash.

**Comment by pzelasko** [comment]
You would be passing in only the base_seed, which is then modified in each node+worker combination similarly to the LM dataset code shared by you earlier. See the snippet linked here. 

https://github.com/lhotse-speech/lhotse/pull/1102/files#diff-4b4592f9922837830d7c5a7fa206aa9c9e8b9c000f9125dc6a651d8d6e1ed446R180

Just make sure to use the IterableDatasetWrapper as described here so that the sampler is being placed into the worker subprocess (so it can correctly read out the worker ID). 

https://github.com/lhotse-speech/lhotse/pull/1102/files#diff-4b4592f9922837830d7c5a7fa206aa9c9e8b9c000f9125dc6a651d8d6e1ed446R30

**Comment by danpovey** [comment]
OK.  I think it's good if we encourage people to use it with IterableDatasetWrapper and also the base_seed (set as a function of at least the start-iter for restarts), because if they don't the use base_seed, in general they'll end up with different base_seed values in different workers, and reproducing failures would be hard.  Debugging isn't something you normally plan in advance to have to do.

**Comment by pzelasko** [comment]
Alright, on another thought I'll simplify this code by removing TRNG and requiring `base_seed` to be passed in by the user.
